OpenRA.Game/Activities/Activity.cs

```diff
@@ -144,15 +144,19 @@ public Activity()
 			IsInterruptible = true;
 		}
 
+		// HACK: Ensure that OnFirstRun always triggers before the first Tick(self), even if Actor.CurrentActivity was updated in a 'dirty' way
+		// that would otherwise skip OnFirstRun (for example through Actor.QueueActivity).
```

Comment:
Wait, what? How does `Actor.QueueActivity` cause skipping `OnFirstRun`?

Reply:
Well, maybe I misinterpreted the cause, but
```
self.CancelActivity();
self.QueueActivity(new ResupplyAircraft(self));
```
results in `OnFirstRun` not triggering (it seems to go right to `Tick` instead, at least before this commit).

Reply:
So I just tested on bleed, and I can't reproduce:

```
Internal mods:
	all: All mods (git-16f1750252)
	cnc: Tiberian Dawn (git-16f1750252)
	d2k: Dune 2000 (git-16f1750252)
	modcontent: Mod Content Manager (git-16f1750252)
	ra: Red Alert (git-16f1750252)
	ts: Tiberian Sun (git-16f1750252)
External mods:
	cnc-git-fd5eaf1f49: Tiberian Dawn (git-fd5eaf1f49)
	d2k-git-09512f0987: Dune 2000 (git-09512f0987)
	ra-playtest-20180729: Red Alert (playtest-20180729)
	ts-git-c1b9df439e: Tiberian Sun (git-c1b9df439e)
	ra-git-16f1750252: Red Alert (git-16f1750252)
Loading mod: ra
System.Net.Sockets.SocketOptionName 0x17 is not supported at IP level
[2019-03-22T21:06:48] Game started
ResupplyAircraft.OnFirstRun()
ResupplyAircraft.Tick()
ResupplyAircraft.Tick()
ResupplyAircraft.Tick()
```

Do I have to follow any special sequence of events, or is ordering an aircraft to an airfield enough?

Reply:
It may not be reproducable on bleed because on bleed, it uses 'normal' queueing instead of ChildActivity, so unless you give the order when already on the pad/field, it has long finished before the aircraft arrives to reload.

To reproduce it on this PR: Revert the last two commits, compile, then in RA let some aircraft deplete some of its ammo, order it to resupply, and as soon as it touches down on the pad/afld and you see the resupply anim starting, press Stop, so the activity is canceled. Now press Stop again, and `ResupplyAircraft.OnFirstRun()` should not trigger, even through the activity was queued after CancelActivity.

Reply:
This is really weird. I poked around a bit and found that `ResupplyActivity` for some arcane reason goes into the `canceling` state before it even gets the chance to tick for the first time. However, the moment the activity is queued it _is_ in the `queued` state as it should. This happens even when there is no previous activity. So somehow something is doing weird things with the activity state somewhere in between those moments.

Reply:
OK, that was stupid.  The reason why that happens is because the Stop order is handled by every trait under the sun, not only `Aircraft`.  In this case, at least `AttackBase` also gets a shot and promptly cancels the activity that was just queued.

The fix would have been to queue the activity in a FrameEndTask.

Reply:
Wait, what? Why is `AttackBase` allowed to cancel non-attack orders?
Fixing that would be trickier, but IMO a better solution using a FrameEndTask workaround.

Reply:
Interesting. I agree that fixing `AttackBase` would be the best solution in the long run, but I don't think it is reasonable to expect this PR to do that. I think the various attack activities need a refactor anyway (e.g. `Attack` is quite misleadingly named IMO).

In the context of what this PR is trying to do, it is not needed. Because the last commit makes the requeueing of `ResupplyActivity` happen on becoming idle, it isn't affected by this bug anymore. You can just drop the second to last commit and everything still works fine.

Reply:
> Why is AttackBase allowed to cancel non-attack orders?

AFAIK we don't check the type of `Actor.CurrentActivity` anywhere where the Stop order gets resolved, we just call `Actor.CancelActivity` without regard to that (which is why it gets called multiple times every time you press S, once from each applicable `ResolveOrder` method). It's only never - until recently - been a problem because we never automatically queued new activities on Stop.

But yes, as tovl mentioned, the FrameEndTask fix is hypothetical since the last commit removes the offending code.  But the second-to-last commit should be removed.

Reply:
> You can just drop the second to last commit and everything still works fine.

Will do that.

About a future proper fix, I think there's several activities we don't want to cancel on "Stop", but we do want them to cancel on pretty much anything else, so maybe we should chain a `bool cancelAttemptByStopOrder` as optional parameter from CancelActivity all the way through to Activity, to allow activities to refuse cancellation by "Stop" orders.

We could also pass through the order string to make things more configurable, but that would be a bit more complicated, and I'm not sure we really need that.

Reply:
> About a future proper fix, I think there's several activities we don't want to cancel on "Stop", but we do want them to cancel on pretty much anything else

What other examples are you thinking of besides `ResupplyAircraft`?

I think the issue Paul is seeing is that traits were stopping an activity unrelated to that trait, not that the activity could be stopped in the first place.

Reply:
> I think the issue Paul is seeing is that traits were stopping an activity unrelated to that trait, not that the activity could be stopped in the first place.

Yeah, exactly.

Re customizing activity stop behaviour: we could implement a `Stop` method on `Activity` with a default implementation that cancels the activity. Special cases can then override this to do other things.

Reply:
> What other examples are you thinking of besides ResupplyAircraft?

`Resupply` (once that PR is merged) for ground units and [Heli]Land, at least. There may be one or two more, but these came to mind first.